### PR TITLE
add addon opa

### PR DESCRIPTION
**Pull request by honeybadger10**

### CODE REVIEW

**Comment by SteveLasker** [comment]
Pinging @ritazh for :eyes:

File: addons/opa/README.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+# create ns
+kubectl create namespace opa
+
+
+# change context to ns opa
+
+
+# generate certs
+openssl genrsa -out ca.key 2048
```
Comment (marcofranssen) [inline addons/opa/README.md:19]: can we move this into a code block like

\`\`\`bash
openssl …………
\`\`\`

File: addons/opa/README.md [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+# create pod and/or deployment
+# for the deployment to succeed, OPA needs to be able to reach the registry
+# change the registry in pod.yaml and deployment.yaml to your registrys IP/FQDN
+kubectl apply -f app/pod.yaml
+kubectl apply -f app/deployment.yaml
+```
+
+
+## denied deployment
```
Comment (marcofranssen) [inline addons/opa/README.md:88]:
```suggestion
## denied deployment

```

File: addons/opa/README.md [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        name: opa
+EOF
+
+# create webhook
+kubectl apply -f webhook-configuration.yaml
+
+```
+
+
+## allowed deployment
```
Comment (marcofranssen) [inline addons/opa/README.md:67]:
```suggestion
## allowed deployment

```

File: addons/opa/README.md [hunk 4]
```diff
@@ -0,0 +1,103 @@
+# nv2-opa
+
+## install OPA
```
Comment (marcofranssen) [inline addons/opa/README.md:3]:
```suggestion
## install OPA

```

**Review by marcofranssen — CHANGES_REQUESTED** [review]
Some minor things captured by markdown linting.

Furthermore I think it would be great to template the k8s resources in a helm chart to have less manual setup steps and replacements in the k8s objects.

It is a good practice to put helm charts in a `charts` folder in the root of the repository.

See here some examples.

https://github.com/philips-labs/helm-charts

See here an example outline to have a quickstart.

```bash
tree charts
charts
└── opa
    ├── Chart.yaml
    ├── LICENSE
    ├── templates
    │   ├── NOTES.txt
    │   ├── _helpers.tpl
    │   ├── admission-controller-configmap.yaml
    │   ├── admission-controller-deployment.yaml
    │   └── admission-controller-service.yaml
    └── values.yaml

2 directories, 8 files
```

You can create this using `helm create` 

```bash
helm create --help

This command creates a chart directory along with the common files and
directories used in a chart.

For example, 'helm create foo' will create a directory structure that looks
something like this:

    foo/
    ├── .helmignore   # Contains patterns to ignore when packaging Helm charts.
    ├── Chart.yaml    # Information about your chart
    ├── values.yaml   # The default values for your templates
    ├── charts/       # Charts that this chart depends on
    └── templates/    # The template files
        └── tests/    # The test files
```

File: addons/opa/rego/nv2.rego [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+
+
+operations := {"CREATE", "UPDATE"}
+kind := {"Deployment", "Pod"}
+
+artifactType := "application/vnd.cncf.notary.v2"
+mediaType := "application/vnd.cncf.notary.signature.v2+jwt"
+
+
+## tbd: how to load certs
```
Comment (ritazh) [inline addons/opa/rego/nv2.rego:12]: Is there a missing step here?

File: addons/opa/rego/nv2.rego [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+}
+
+## get nv2 signature
+get_nv2_jwt(image_obj, digest) = nv2_jwt {
+  
+	url_digest_arr := [ "https:/", image_obj.registry, "v2", image_obj.image, "blobs", digest ]
+  url_digest := concat("/", url_digest_arr)
+  
+	headers_json := {"Content-Type": "application/json"}
+  response_signature := http.send({"method": "get", "url": url_digest, "headers": headers_json, "tls_insecure_skip_verify": true})
```
Comment (ritazh) [inline addons/opa/rego/nv2.rego:165]: can you give an example of an url that returns a signature?

**Comment by sozercan** [comment]
does this work with private registries?

**Comment by ritazh** [comment]
In general, it would be helpful to include instructions for:
- prerequisites for which registry supports this feature and how to enable the feature on the registry
- setting up nv2

**Comment by SteveLasker** [comment]
> does this work with private registries? or public ones that requires tokens?

Oh, that's a good point. CNCF distribution doesn't have auth, but any cloud deployment would need to validate with an authenticated connection to the registry. Most cloud providers would have a Service Identity like configuration, so I'm not sure how that associates with OPA/Gatekeeper as customers are really trying to move away from any secrets model as it complicates key rotation.

**Comment by SteveLasker** [comment]
> In general, it would be helpful to include instructions for:
> * prerequisites for which registry supports this feature and how to enable the feature on the registry

At this point, nobody supports it as we're still incubating the idea, so only a local instance of CNCF distribution will work. 
Fall 2021, that's another story :)

**Comment by SteveLasker** [comment]
Adding @mnltejaswini ^

**Comment by SteveLasker** [comment]
@honeybadger10, just a ping. We haven't forgotten about this. There's some work happening to enable data providers to enable the passthrough of data within OPA to enable the flows.  
We should be able to pop this up again soon.

**Comment by dtzar** [comment]
@honeybadger10 - thank you for your PR.  It's been inactive for a long time and we want to avoid specific implementations which utilize notation in this repository, so I'm closing it out.  Perhaps you could put something into your own repository?

For instance, we have [ratify](https://github.com/deislabs/ratify) which utilizes notation v2 with gatekeeper and there are others like [connaisseur](https://github.com/sse-secure-systems/connaisseur) - currently only supports notary v1.

**Pull request closed without merging**